src/qibolab/_core/instruments/keysight/sweep.py

```diff
@@ -0,0 +1,110 @@
+        to be swept and corresponding QCS variable.
+    """
+    hardware_sweepers: list[tuple[list[qcs.Array], list[qcs.Scalar]]] = []
+    software_sweepers: list[tuple[list[qcs.Array], list[qcs.Scalar]]] = []
+    hw_sweeper_order = []
+    sw_sweeper_order = []
+
+    # Mapper for pulses that are controlled by a sweeper and the parameter to be swept
+    sweeper_pulse_map: defaultdict[PulseId, dict[str, qcs.Scalar]] = defaultdict(dict)
+    # Mapper for channels with frequency controlled by a sweeper
+    sweeper_channel_map: dict[ChannelId, qcs.Scalar] = {}
+
+    for idx, parallel_sweeper in enumerate(sweepers):
+        sweep_values: list[qcs.Array] = []
+        sweep_variables: list[qcs.Variable] = []
+        # Currently nested hardware sweeping is not supported
+        hardware_sweeping = len(hardware_sweepers) == 0
+
+        for idx2, sweeper in enumerate(parallel_sweeper):
+            qcs_variable = qcs.Scalar(
+                name=f"V{idx}_{idx2}", value=sweeper.values[0], dtype=float
+            )
+
+            if sweeper.parameter is Parameter.frequency:
+                sweeper_channel_map.update(
+                    {channel_id: qcs_variable for channel_id in sweeper.channels}
+                )
+                # Readout frequency is not supported with hardware sweeping
+                if not probe_channel_ids.isdisjoint(sweeper.channels):
+                    hardware_sweeping = False
+            elif sweeper.parameter in [
+                Parameter.amplitude,
+                Parameter.duration,
+                Parameter.relative_phase,
+            ]:
+                # Duration is not supported with hardware sweeping
+                if sweeper.parameter is Parameter.duration:
+                    hardware_sweeping = False
+
+                for pulse in sweeper.pulses:
+                    sweeper_pulse_map[pulse.id][sweeper.parameter.name] = qcs_variable
+            else:
+                raise ValueError(
+                    "Sweeper parameter not supported", sweeper.parameter.name
+                )
+
+            sweep_variables.append(qcs_variable)
+            sweep_values.append(
+                qcs.Array(
+                    name=f"A{idx}_{idx2}",
+                    value=(
+                        sweeper.values * NS_TO_S
+                        if sweeper.parameter is Parameter.duration
+                        else sweeper.values
+                    ),
+                    dtype=float,
+                )
+            )
+        if hardware_sweeping:
+            hardware_sweepers.append((sweep_values, sweep_variables))
+            hw_sweeper_order.append(idx)
+        else:
+            software_sweepers.append((sweep_values, sweep_variables))
+            sw_sweeper_order.append(idx)
+
+    sweeper_swaps_required = [
+        (original_index, shifted_index)
+        for shifted_index, original_index in enumerate(
+            hw_sweeper_order + sw_sweeper_order
```

Comment:
So the usecase I was thinking about is that if we have hardware-software-hardware, and we can get a bit of optimization doing hardware-hardware-software.

I guess the current motivation would be just optimizing currently used 2D scans and having the hardware sweeper first.

Reply:
Ok, this has the potential of being complex and confusing. But it's not a problem of Keysight alone.

On the one side, I would propose to avoid sorting the sweepers, and just honor what the user specified in the `Platform.execute()` call as it is.
However, it's true that there are applications that are written in a hardware-agnostic way (Qibocal in the first place), which may benefit from some automated hardware-aware optimization, since you don't really care about the sweepers sorting, but you don't even know which is the most optimal one ahead of time, since it may depend on the driver.

This could be an extension to #854, in which we add a flag for `sweepers_auto_sorting` in the `ExecutionParameters` (or something like that - it's just a random name), and Qibolab itself could optimize the sorting based on which sweepers are implemented on hardware by the chosen driver.

In any case, this is not happening tomorrow, thus feel free to decide which behavior you prefer to support (honoring the input or always auto-sorting).

Reply:
Yeah, I'm leaning towards honouring the user calls as they are. I'll revert it back to the original approach and test tomorrow.

Reply:
Changed in https://github.com/qiboteam/qibolab/pull/944/commits/241de1acd098c3483fba8a4fe63b586bf9a186d2